modules/imgproc/src/opencl/pyramid_up.cl

```diff
@@ -0,0 +1,125 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+__constant float kx[] = { 0.125, 0.5, 0.75, 0.5, 0.125 };
+__constant float ky[] = { 0.125, 0.5, 0.75, 0.5, 0.125 };
+
+#define OP(delta, y, x) (convert_float4(arr[(y + delta) * 5 + x]) * ky[y] * kx[x])
+
+__kernel void pyrUp_cols2(__global const uchar * src, int src_step, int src_offset, int src_rows, int src_cols,
+                          __global uchar * dst, int dst_step, int dst_offset, int dst_rows, int dst_cols)
+{
+    int block_x = get_global_id(0);
+    int y = get_global_id(1) * 2;
+    int ssx, dsx;
+
+    if ((block_x * 4) >= dst_cols || y >= dst_rows) return;
+
+    uchar8 line[6];
+    uchar4 line_out;
+
+    ssx = dsx = 1;
+    int offset, src_index;
+    src_index = block_x * 2 * ssx + (y / 2 - 2) * src_step - 1 + src_offset;
```

Comment:
actually not src_index,  but "src_index + xxx * src_step" should be checked negative or not.  which is the offset of line[0] and line[2] in the kernel.  line[0] and line[2] offset can be both negative or line[0] offset can be negative and line[2] offset be positive. so I checked them individually here.